flow/StreamCipher.cpp

```diff
@@ -19,75 +19,83 @@
+#include "flow/ITrace.h"
 #include "flow/UnitTest.h"
+#include <memory>
 
 std::unordered_set<EVP_CIPHER_CTX*> StreamCipher::ctxs;
-std::unique_ptr<StreamCipher::Key> StreamCipher::Key::globalKey;
+std::unique_ptr<StreamCipherKey> StreamCipherKey::globalKey;
 
-StreamCipher::StreamCipher() : ctx(EVP_CIPHER_CTX_new()) {
-	ctxs.insert(ctx);
+StreamCipherKey* StreamCipherKey::getGlobalCipherKey() {
+	if (StreamCipherKey::globalKey.get() == nullptr) {
+		globalKey = std::make_unique<StreamCipherKey>(AES_256_KEY_LENGTH);
+		return globalKey.get();
+	}
+	return globalKey.get();
 }
 
-StreamCipher::~StreamCipher() {
-	EVP_CIPHER_CTX_free(ctx);
-	ctxs.erase(ctx);
+void StreamCipherKey::setGlobalCipherKey(uint8_t* const data, int len) {
+	StreamCipherKey* cipherKey = StreamCipherKey::getGlobalCipherKey();
+	cipherKey->initializeKey(data, len);
 }
 
-EVP_CIPHER_CTX* StreamCipher::getCtx() {
-	return ctx;
-}
-
-void StreamCipher::cleanup() noexcept {
-	Key::cleanup();
-	for (auto ctx : ctxs) {
-		EVP_CIPHER_CTX_free(ctx);
+void StreamCipherKey::globalKeyCleanup() {
+	if (StreamCipherKey::globalKey != nullptr) {
+		StreamCipherKey* cipherKey = StreamCipherKey::getGlobalCipherKey();
+		cipherKey->cleanup();
 	}
 }
 
-void StreamCipher::Key::initializeKey(RawKeyType&& arr) {
-	if (globalKey) {
-		ASSERT(globalKey->arr == arr);
-	}
-	globalKey = std::make_unique<Key>(ConstructorTag{});
-	globalKey->arr = std::move(arr);
-	memset(arr.data(), 0, arr.size());
+StreamCipher::StreamCipher(int keySize)
+  : ctx(EVP_CIPHER_CTX_new()), hmacCtx(HMAC_CTX_new()), cipherKey(std::make_unique<StreamCipherKey>(keySize)) {
+	ctxs.insert(ctx);
 }
 
-void StreamCipher::Key::initializeRandomTestKey() {
-	ASSERT(g_network->isSimulated());
-	if (globalKey)
-		return;
-	globalKey = std::make_unique<Key>(ConstructorTag{});
-	generateRandomData(globalKey->arr.data(), globalKey->arr.size());
+StreamCipher::StreamCipher()
+  : ctx(EVP_CIPHER_CTX_new()), hmacCtx(HMAC_CTX_new()),
+    cipherKey(std::make_unique<StreamCipherKey>(AES_256_KEY_LENGTH)) {
+	ctxs.insert(ctx);
 }
 
-const StreamCipher::Key& StreamCipher::Key::getKey() {
-	ASSERT(globalKey);
-	return *globalKey;
+StreamCipher::~StreamCipher() {
+	HMAC_CTX_free(hmacCtx);
+	EVP_CIPHER_CTX_free(ctx);
+	ctxs.erase(ctx);
 }
 
-StreamCipher::Key::Key(Key&& rhs) : arr(std::move(rhs.arr)) {
-	memset(arr.data(), 0, arr.size());
+EVP_CIPHER_CTX* StreamCipher::getCtx() {
+	return ctx;
 }
 
-StreamCipher::Key& StreamCipher::Key::operator=(Key&& rhs) {
-	arr = std::move(rhs.arr);
-	memset(arr.data(), 0, arr.size());
-	return *this;
+HMAC_CTX* StreamCipher::getHmacCtx() {
+	return hmacCtx;
 }
 
-StreamCipher::Key::~Key() {
-	memset(arr.data(), 0, arr.size());
+void StreamCipher::cleanup() noexcept {
+	StreamCipherKey::globalKeyCleanup();
+	for (auto ctx : ctxs) {
+		EVP_CIPHER_CTX_free(ctx);
+	}
 }
 
-void StreamCipher::Key::cleanup() noexcept {
-	globalKey.reset();
+void applyHmacKeyDerivationFunc(StreamCipherKey* const cipherKey,
```

Comment:
Is this meant to be `StreamCipherKey const *`?

Reply:
Typo, should fix it.